File: 2/PROLOG/ackermann_phi.pl
:- use_module(library(statistics)).

ackermann_phi(M, N, 0, Result) :-
    Result is M + N.
ackermann_phi(_, _, 1, 0).
ackermann_phi(_, _, 2, 1).
ackermann_phi(M, _, 3, M).
ackermann_phi(M, 0, P, Result) :-
    P > 2,
    P1 is P - 1,
    ackermann_phi(M, 0, P1, Result).
ackermann_phi(M, N, P, Result) :-
    N > 0,
    P > 0,
    N1 is N - 1,
    P1 is P - 1,
    ackermann_phi(M, N1, P, TempResult),
    ackermann_phi(M, TempResult, P1, Result).

compute_phi :-
    between(0, 3, M),
    between(0, 3, N),
    between(0, 4, P),
    time((ackermann_phi(M, N, P, Result), write('φ('), write(M), write(', '), write(N), write(', '), write(P), write(') = '), write(Result), nl)),
    fail.
compute_phi.

File: 3/C ANSI/ackermann_phi.c
#include <stdio.h>

int ackermann_phi(int m, int n, int p) {
    if (p == 0) return m + n;
    if (p == 1) return 0;
    if (p == 2) return 1;
    if (p == 3) return m;
    if (n == 0 && p > 2) return ackermann_phi(m, 0, p - 1);
    return ackermann_phi(m, ackermann_phi(m, n - 1, p), p - 1);
}

int main() {
    int m = 2, n = 3, p = 4;
    printf("φ(%d, %d, %d) = %d\n", m, n, p, ackermann_phi(m, n, p));
    return 0;
}
